internal/services/shared/errors.go

```diff
@@ -66,6 +69,43 @@ func (err ErrSchemaWriteDataValidation) GRPCStatus() *status.Status {
 	)
 }
 
+// MaxDepthExceededError is an error returned when the maximum depth for dispatching has been exceeded.
+type MaxDepthExceededError struct {
+	error
+
+	// AllowedMaximumDepth is the configured allowed maximum depth.
+	AllowedMaximumDepth uint32
+}
+
+// GRPCStatus implements retrieving the gRPC status for the error.
+func (err MaxDepthExceededError) GRPCStatus() *status.Status {
+	return spiceerrors.WithCodeAndDetails(
+		err,
+		codes.ResourceExhausted,
+		spiceerrors.ForReason(
+			v1.ErrorReason_ERROR_REASON_MAXIMUM_DEPTH_EXCEEDED,
+			map[string]string{
+				"maximum_depth_allowed": strconv.Itoa(int(err.AllowedMaximumDepth)),
+			},
+		),
+	)
+}
+
+// NewMaxDepthExceededError creates a new MaxDepthExceededError.
+func NewMaxDepthExceededError(allowedMaximumDepth uint32, isCheckRequest bool) error {
+	if isCheckRequest {
+		return MaxDepthExceededError{
+			fmt.Errorf("the check request has exceeded the allowable maximum depth of %d: this usually indicates a recursive or too deep data dependency. Try running zed with --explain to see the dependency. See: https://spicedb.dev/d/debug-max-depth-check", allowedMaximumDepth),
```

Comment:
I think that instead of putting the "suggestion" or "recommendation" in the error message, we could put it as a new field in the error metadata. That way error can be what the error is, and then there can be 1 or more recommendations to look into as extra metadata

Reply:
I disagree, only because the error message is what most people are going to read, and debugging the issue is going to be top of their mind